RelationalAI/DefaultAccessTokenHandler.cs

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2022 RelationalAI, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+
+namespace RelationalAI
+{
+    // This handler caches tokens in ~/.rai/tokens.json. It will attempt to load
+    // a token from the cache file and if it is not found or has expired, it will
+    // delegate to rest.RequestAccessTokenAsync to retrieve a new token and will save it
+    // in the cache file.
+    public class DefaultAccessTokenHandler : IAccessTokenHandler
+    {
+        private static readonly SemaphoreSlim SemaphoreSlim = new SemaphoreSlim(1);
+
+        public Rest Rest { get; set; }
```

Comment:
remove these public REST/Logger properties after fixing the interface as described in the interface